openssl/src/x509/mod.rs

```diff
@@ -443,6 +460,153 @@ impl X509Ref {
+                ptr::null_mut(),
+                ptr::null_mut(),
+            );
+
+            if ext_key_usage.is_null() {
+                None
+            } else {
+                Some(Stack::from_ptr(ext_key_usage as *mut _))
+            }
+        }
+    }
+
+    /// Returns this certificate `AuthorityKeyId`, if it exists.
+    ///
+    /// This corresponds to [`X509_get_ext_d2i`] called with `NID_authority_key_identifier`.
+    ///
+    /// [`X509_get_ext_d2i`]: https://www.openssl.org/docs/man1.1.0/crypto/X509_get_ext_d2i.html
+    pub fn authority_keyid(&self) -> Option<AuthorityKeyId> {
+        unsafe {
+            let akid = ffi::X509_get_ext_d2i(
+                self.as_ptr(),
+                ffi::NID_authority_key_identifier,
+                ptr::null_mut(),
+                ptr::null_mut(),
+            );
+
+            if akid.is_null() {
+                None
+            } else {
+                Some(AuthorityKeyId::from_ptr(akid as *mut _))
+            }
+        }
+    }
+
+    /// Returns this certificate subject key ID, if it exists.
+    ///
+    /// This corresponds to [`X509_get_ext_d2i`] called with `NID_subject_key_identifier`.
+    ///
+    /// [`X509_get_ext_d2i`]: https://www.openssl.org/docs/man1.1.0/crypto/X509_get_ext_d2i.html
+    pub fn subject_keyid(&self) -> Option<Asn1String> {
+        unsafe {
+            let skid = ffi::X509_get_ext_d2i(
+                self.as_ptr(),
+                ffi::NID_subject_key_identifier,
+                ptr::null_mut(),
+                ptr::null_mut(),
+            );
+
+            if skid.is_null() {
+                None
+            } else {
+                Some(Asn1String::from_ptr(skid as *mut _))
+            }
+        }
+    }
+
+    /// Returns an iterator that allows iterating over the `X509Extensions` present in
+    /// this certificate.
+    ///
+    /// This issues one call to [`X509_get_ext_count`] to get the number of extensions, while the
+    /// iterator itself will lazily fetch each extension.
+    ///
+    /// [`X509_get_ext_count`]: https://www.openssl.org/docs/man1.1.0/crypto/X509_get_ext_count.html
+    pub fn extensions(&self) -> Extensions {
+        let extension_count: c_int;
+
+        unsafe {
+            extension_count = ffi::X509_get_ext_count(self.as_ptr());
+        }
+
+        Extensions {
+            idxs: 0..extension_count,
+            cert: &*self,
+        }
+    }
+
+    /// Returns an `X509ExtensionRef` corresponding to the given NID, if it exists.
+    ///
+    /// This is corresponds to [`X509_get_by_NID`] to get the NID position, followed by a call
+    /// to [`X509_get_ext`] if the extension exists.
+    ///
+    /// [`X509_get_by_NID`]: https://www.openssl.org/docs/man1.1.0/crypto/X509_get_ext_by_NID.html
+    /// [`X509_get_ext`]: https://www.openssl.org/docs/man1.1.0/crypto/X509_get_ext.html
+    pub fn extension_from_nid(&self, nid: Nid) -> Option<&X509ExtensionRef> {
```

Comment:
Let's call this `extension_by_nid` to match a bit more closely with the OpenSSL naming.